crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1055,36 +1055,62 @@ impl OlmMachine {
-            // key that sent us the session (recorded upon successful
-            // decryption of the `m.room_key` to-device message).
-            if let Some(device) = self
-                .get_user_devices(sender, None)
-                .await?
-                .devices()
-                .find(|d| d.curve25519_key() == Some(session.sender_key()))
-            {
-                // If the `Device` is confirmed to be the owner of the
-                // `InboundGroupSession` we will consider the session (i.e.
-                // "room key"), and by extension any events that are encrypted
-                // using this session, trusted if either:
-                //
-                //     a) This is our own device, or
-                //     b) The device itself is considered to be trusted.
-                if device.is_owner_of_session(session)
-                    && (device.is_our_own_device() || device.is_verified())
-                {
-                    (VerificationState::Trusted, Some(device.device_id().to_owned()))
+        Ok(if session.has_been_imported() {
+            // the sender_key is claimed we can't check any authenticity
+            // No point of trying to find the matching device
+            (VerificationState::UnsafeSource, None)
+        } else if let Some(device) = self
+            .get_user_devices(sender, None)
+            .await?
+            .devices()
+            .find(|d| d.curve25519_key() == Some(session.sender_key()))
+        {
+            // We found a matching device, let's check if it owns the session
+            if device.is_owner_of_session(session) {
+                // We want to make a distinction between an unsigned device of a verified
+                // owner and an unsigned device of an unverified user
+                let device_owner_verified = device.is_device_owner_verified();
+
+                if device.is_our_own_device() || device.is_verified() {
+                    // Special case that to consider local trust
+                    (VerificationState::Verified, Some(device.device_id().to_owned()))
+                } else if device.is_cross_signed_by_owner() {
+                    // The device is cross signed by this owner
+                    // Meaning that the user did self verify it properly
+                    // Let's check if we trust the identity
+                    if device_owner_verified {
+                        // can this happen? in this case the device would be verified
+                        (VerificationState::Verified, Some(device.device_id().to_owned()))
+                    } else {
+                        (
+                            VerificationState::SignedDeviceOfUnverifiedUser,
+                            Some(device.device_id().to_owned()),
+                        )
+                    }
+                } else if device_owner_verified {
+                    // The user is verified by us but he hasn't self verified his device
```

Comment:
```suggestion
                    // The user is verified by us but they haven't self verified his device
```

or

```suggestion
                    // The user is verified by us but the user hasn't self verified his device
```

Just trying to be neutral :-/

Reply:
In that case...

```suggestion
                    // The user is verified by us but the user hasn't self-verified their device
```